### PR TITLE
Convert adapters to Python

### PR DESCRIPTION
**Pull request by evacchi**
While [setting up CI on GHA for wazero](https://github.com/tetratelabs/wazero/pull/1062) to run against the test suite, we realized that the Python script is assuming a Unix environment. I am opening this PR to start a discussion on the best approach. In my own tests, this worked reliably.

This PR:

1. allows `colorama>=0.4.3`, otherwise pip on Windows (GHA `windows-2022`) raises the error:

```
INFO: pip is looking at multiple versions of colorama to determine which version is compatible with other requirements. This could take a while.
ERROR: Cannot install -r requirements/dev.txt (line 4) and colorama==0.4.3 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested colorama==0.4.3
    pylint 2.14.3 depends on colorama>=0.4.5; sys_platform == "win32"
```

2. converts the shell-based adapters to Python and invokes them through the current Python interpreter (in a new sub-process)

3. updates docs and GitHub actions accordingly

<details> <summary>(Originally proposed changes – no longer apply)</summary>
2. we explicitly invoke the shell adapter through `bash` instead of relying on the `#!` directive, because on Windows that won't work. Of course this also means that all adapters would be now expected to be runnable on bash.

4. as a further degree of complexity, it looks like proper `bash` (installed under `C:\Program Files\Git\bin` is not on the PATH when invoked from Python on GHA, invoking instead `C:\Windows\system32\bash.exe` which is just a stub that prints a message telling the user to install WSL (which of course we don't want as that's a virtualized linux distro).

~Instead of fiddling with PATH, I decided to export an env variable (`TEST_SHELL_RUNNER`), which defaults to `bash` that can be overridden to a full path to a shell executable.~

~EDIT: minimal reproducer https://github.com/evacchi/gha-playground/actions/runs/4013516139/jobs/6892946727~

EDIT2: I found a better workaround https://github.com/actions/runner-images/pull/1081#issuecomment-661524144

Signed-off-by: Edoardo Vacchi <evacchi@users.noreply.github.com>
</details>

### CODE REVIEW

**Comment by loganek** [comment]
Thanks a lot for the change. I do understand the motivation and I think it makes sense for Github Workflow. However, this change adds a requirement that all the adapters now have to be bash scripts. I'd rather avoid that unless it's absolutely necessary. Could we potentially solve it in a way so it doesn't force adapter maintainers to only write bash scripts?

**Comment by evacchi** [comment]
Obviously fair point. I have thought of two possible solutions. Neither is great 😅

1) parse the shebang. Kind of a hack. Moreover either we expect the shebang to be of the form 

    !#bash

i.e. a bare executable without leading path, or we'll also need to parse the path (of course /bin/sh won't make any sense on windows)

2) add a tiny config file next to the adapter. While this feels less hackish it is basically reproducing the shebang, including the path issue

While I would favor (2) it's only slightly less of a hack

**Comment by codefromthecrypt** [comment]
PS after this change, there's still a problem with the shell scripts (at least in git bash). For example, assemblyscript `environ_get-multiple-variables` fails when run in anything including wasmtime due to the parse arg loop not handling newlines properly.

*It may be a better choice to not use bash and also python, rather just python?* In other words maybe we can avoid the whole fork bash problems by not using it and instead have the shims python?

```bash
$ python3 test-runner/wasi_test_runner.py -t ./tests/assemblyscript/testsuite/ -r adapters/wasmtime.sh
--snip--
Test environ_get-multiple-variables failed
  [exit_code] 0 == 1
STDOUT
Unknown option line
--snip--
```

**Comment by evacchi** [comment]
using Python would seem fair to me. After all the adapters are pretty simple

**Comment by loganek** [comment]
I'd be ok with using Python as adapter, given it's very simple and we already require it for runtime anyway. @evacchi would you like to update your PR to support Python for adapters?

**Comment by evacchi** [comment]
> I'd be ok with using Python as adapter, given it's very simple and we already require it for runtime anyway. @evacchi would you like to update your PR to support Python for adapters?

so translate the other adapters? sure that's doable if you're ok with it :)

File: adapters/wasm-micro-runtime.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+parser.add_argument('--version', action='store_true')
+parser.add_argument('--test-file', action='store')
+parser.add_argument('--arg', action='append', default=[])
+parser.add_argument('--env', action='append', default=[])
+parser.add_argument('--dir', action='append', default=[])
+
+args = parser.parse_args()
+
+if args.version:
+    subprocess.run(['iwasm', '--version'])
```
Comment (loganek) [inline adapters/wasm-micro-runtime.py:15]: we've recently updated adapters so custom path to executable can be provided. Could we add that change here too?

**Review by loganek — CHANGES_REQUESTED** [review]
Thanks a lot, the change looks good to me. I left a few minor comments. i think we should also delete bash adapters once we have python ones too.

I think we should also update the documentation and and CI jobs. Please let me know if you can help with those as well.

File: adapters/runtime_adapter.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+import subprocess
```
[the remark this answers is not in the document]
Comment (evacchi) [reply adapters/runtime_adapter.py:1]: hmm, probably something went wrong while merging

**Comment by evacchi** [comment]
I think I have applied the changes you suggested; obviously feel free to add more comments :) in case all is good, feel also free to squash the PR and merge (unless we want me to rebase and squash explicitly)

Thanks!

**Review by loganek — APPROVED** [review]
The change looks good. I don't know of any other runtime that implements the adapter (other than toywasm, which already is written in Python), so I think we can merge it now.

Many thanks for making the change!

**Pull request merged**